### PR TITLE
Use workspace inheritance to specify versions of dependencies

### PR DESCRIPTION
**Pull request by mina86**
Move all dependencies to root Cargo.toml file and refer to it using
inheritance in all members of the workspace.  This will hopefully help
reduce noise when updating a package in the future.  The only
exception is `rand` dependency in core/crypto crate which needs to be
held back at older version because of ed25519-dalek.

Doing this, put union on cargo features in the root Cargo.toml so
features don’t need to be enabled in member crates.  The one exception
to this rule is serde_json’s preserve_order feature which is only
enabled in tools that need it.

Fixes: https://github.com/near/nearcore/issues/6301
Fixes: https://github.com/near/nearcore/issues/7674

### CODE REVIEW

**Comment by mina86** [comment]
This does bring a few questions:
1. Should we do this at all?
2. Should we do this for all dependencies?
3. Should we use union of features in workspace Cargo.toml and then
   not specify further features in members’ Cargo.toml files?

**Comment by akhi3030** [comment]
Fixes https://github.com/near/nearcore/issues/7674

**Comment by akhi3030** [comment]
>     1. Should we do this at all?

What are the pros and cons?

**Comment by mina86** [comment]
> > 1. Should we do this at all?
> 
> What are the pros and cons?

Pro is of course that we don’t need to specify version in multiple places.  I’m not sure of cons.

**Comment by akhi3030** [comment]
> > > 1. Should we do this at all?
> > 
> > 
> > What are the pros and cons?
> 
> Pro is of course that we don’t need to specify version in multiple places. I’m not sure of cons.

Indeed, I am also not aware of any cons.  @nagisa: maybe you have thoughts on this as well?

File: chain/chain/Cargo.toml [hunk 1]
```diff
@@ -8,21 +8,21 @@ rust-version.workspace = true
...
+ansi_term.workspace = true
+borsh.workspace = true
+chrono = { workspace = true, features = ["serde"] }
+crossbeam-channel.workspace = true
+enum-map.workspace = true
+itertools.workspace = true
+lru.workspace = true
+num-rational.workspace = true
+once_cell.workspace = true
 rand = "0.7"
```
Comment (akhi3030) [inline chain/chain/Cargo.toml:21]: If we decide to move forward with this PR, do we plan on doing subsequent PRs to synchronise versions for the crates that could not be moved in this PR.

**Comment by matklad** [comment]
1) I think we should -- duplicating versions across Cargo.tomls is not fun. For some specific cases, I'd say that re-export (eg, `near_o11y::prometheus`) might better (as you also save on the *number* of deps in Cargo.tomls), but for stuff like rocks or serde, pulling them up to workspace level seems like a win
2) I am not sure. I can see "only if used at least twice" also being a reasonable rule. We could also try this out inceremntally, by initially moving only known painful cases (rocks, tracing) and seeing how it goes. OTOH, "only declare deps in the root cargo.toml" is very simple, and easy to keep consistent. My vote is probably for "let's keep all deps in root". 
3) Fairly certain that it's better to only enable features in the root, otherwise Cargo is very prone to spurious recompiles dependenig on which subset of packages is being built.

**Comment by nagisa** [comment]
I’m not aware of any specific cons either. This is a very new feature and so we’d be the ones to discover them were we land this change soon. I’m also of the mind that if we’re using this feature we should keep all the versions at the top-level `Cargo.toml` so that exceptions don’t become necessary (“you can see all versions in top-level Cargo.toml, except for… this, and that, which can be found here, or there”)

File: chain/chain/Cargo.toml (hunk 1 above)
Comment (mina86) [reply chain/chain/Cargo.toml:21]: I’m working on a change to move all rand to 0.8.5.

Comment (mina86) [reply chain/chain/Cargo.toml:21]: https://github.com/near/nearcore/pull/7705

Comment (mina86) [reply chain/chain/Cargo.toml:21]: Either works for me.

**Review by matklad — APPROVED** [review]
Seems we have a rough consensus here!

**Review by akhi3030 — APPROVED** [review]

**Pull request merged**